### ISSUE TITLE
Make Plonky3 Keccak machine work with memory pointers

### ISSUE THREAD

**Issue by qwang98** [powdr-labs/powdr#1803]
Currently the operation uses 200 assignment registers. Need to use memory addresses of input/output arrays instead, similar to Poseidon.

**Comment by leonardoalt** [comment]
Oh, @qwang98 I think this should be done first.

**Comment by qwang98** [comment]
@georgwiese Would appreciate some advice on solution number 3 above.

**Comment by chriseth** [comment]
Can you explain a bit more how the data flow in your current machine works? Would the problem be solved for now by adding a fourth pass (which is trivial)? Is witgen incorrect currently (I don't fully understand what you mean by "still need to write to memory")?

Witgen for many block machines is inefficient because of the data flow. Once we focus on fast auto-witgen, the system could be analyzed and generate code that directly derives code for the first row from the latch row.

**Comment by georgwiese** [comment]
Yes, like Chris said, these witgen passes are just a heuristic and not set in stone. If we can fix it by changing those, we should do that instead of changing the machine to fit into the current heuristics! I think that rules out option 2. (I actually think we currently do [backward, forward, backward, forward](https://github.com/powdr-labs/powdr/blob/8595b4e54a5160859ff30713cf546b95bd7cca46/executor/src/witgen/sequence_iterator.rs#L44-L47), so 4 passes).

I think I understand the data flow, so let me try to paraphrase the problem: The "core" Keccak machine needs all 50 input words in the first row and has all 50 output words in the last row. This is very similar to the [`PoseidonGLMemory` machine](https://github.com/powdr-labs/powdr/blob/8595b4e54a5160859ff30713cf546b95bd7cca46/std/machines/hash/poseidon_gl_memory.asm). There are two solutions I'm aware of:
1. `PoseidonGLMemory` has a copy of all inputs and outputs and keeps them constant in a block. In your case, this would amount to 200 columns (2 limbs per word * 50 * 2 (input and output)). The existing `Keccakf16` machine already [has that for the input](https://github.com/powdr-labs/powdr/blob/8595b4e54a5160859ff30713cf546b95bd7cca46/std/machines/hash/keccakf16.asm#L152). Then, we'd have relatively few links to spread all the memory reads / write across the 24 rows of the block. I think this solution is similar to yours, but I don't understand how you get away with only 12 extra columns! I think you'd have to add another 100 columns to repeat the output in all rows of the block, so that they can be written to memory in different rows. Or maybe these 100 columns can be shared somehow (e.g. they'd contain the inputs in the first half and the outputs in the second half). 
2. Do all reads in the first row and all writes in the last row. This is your solution 1 and also happens to be what [SP1 is doing](https://github.com/succinctlabs/sp1/blob/4cac73b4692232a964d4509eaf8a856b88a65776/crates/core/machine/src/syscall/precompiles/keccak256/air.rs).

Thinking about it, I'd actually think that solution 2 is cheaper, because then we don't need to repeat the inputs and outputs in all rows! The cost is many more links (which are inactive in all rows except the first and the last). It also seems easier to implement and witgen would probably be easier to follow ^^

**Comment by qwang98** [comment]
Thanks both for the comments. Sorry should have linked my current solution here: https://github.com/powdr-labs/powdr/pull/1832/files#diff-f9fcf1597f4c7e16da854fd1ded4cbcfed738cc44655382ec0cec5140701f8e9

It's actually 139 more columns. 12 is just for the addresses.

**Comment by qwang98** [comment]
> Can you explain a bit more how the data flow in your current machine works? Would the problem be solved for now by adding a fourth pass (which is trivial)? Is witgen incorrect currently (I don't fully understand what you mean by "still need to write to memory")?
> 
> Witgen for many block machines is inefficient because of the data flow. Once we focus on fast auto-witgen, the system could be analyzed and generate code that directly derives code for the first row from the latch row.

My machine is 24 rows and does keccak computation from top to bottom, so the work flow is basically reading inputs from top to bottom, then a wasted pass from bottom to top, then keccak computation from top to bottom, and will need another pass to write output from bottom to top.

**Comment by qwang98** [comment]
@georgwiese Seems to me (from just the literal reading of it) that this is forward, backward, forward, backward? https://github.com/powdr-labs/powdr/blob/8595b4e54a5160859ff30713cf546b95bd7cca46/executor/src/witgen/sequence_iterator.rs#L44-L47

**Comment by qwang98** [comment]
@georgwiese I think your suggestion makes sense. I'd do my solution 1 originally, i.e. doing many links in the first and last step.

**Comment by georgwiese** [comment]
For consistency with the other "precompile" issues, I'll add the tasks:

- [ ] Add machine & test to standard library (WIP: #1832)
- [ ] Integrate end-to-end via syscall in RISC-V machine

**Comment by leonardoalt** [comment]
This is done.
